core/arch/arm/pta/core_fs_htree_tests.c

```diff
@@ -0,0 +1,409 @@
+	size_t offs = op->params[0].u.value.b;
+	size_t sz = op->params[0].u.value.c;
+	size_t end = offs + sz;
+
+	if (end > a->data_alloced) {
+		EMSG("out of bounds");
+		return TEE_ERROR_GENERIC;
+	}
+
+	memcpy(a->data + offs, a->block, sz);
+	if (end > a->data_len)
+		a->data_len = end;
+	return TEE_SUCCESS;
+
+}
+
+static const struct tee_fs_htree_storage test_htree_ops = {
+	.block_size = TEST_BLOCK_SIZE,
+	.rpc_read_init = test_read_init,
+	.rpc_read_final = test_read_final,
+	.rpc_write_init = test_write_init,
+	.rpc_write_final = test_write_final,
+};
+
+#define CHECK_RES(res, cleanup)						\
+		do {							\
+			TEE_Result _res = (res);			\
+									\
+			if ((res)) {					\
+				EMSG("error: res = %#" PRIx32, _res);	\
+				{ cleanup; }				\
+			}						\
+		} while (0)
+
+static uint32_t val_from_bn_n_salt(size_t bn, size_t n, uint8_t salt)
+{
+	assert(bn < UINT16_MAX);
+	assert(n < UINT8_MAX);
+	return SHIFT_U32(n, 16) | SHIFT_U32(bn, 8) | salt;
+}
+
+static TEE_Result write_block(struct tee_fs_htree **ht, size_t bn, uint8_t salt)
+{
+	uint32_t b[TEST_BLOCK_SIZE / sizeof(uint32_t)];
+	size_t n;
+
+	for (n = 0; n < ARRAY_SIZE(b); n++)
+		b[n] = val_from_bn_n_salt(bn, n, salt);
+
+	return tee_fs_htree_write_block(ht, bn, b);
+}
+
+static TEE_Result read_block(struct tee_fs_htree **ht, size_t bn, uint8_t salt)
+{
+	TEE_Result res;
+	uint32_t b[TEST_BLOCK_SIZE / sizeof(uint32_t)];
+	size_t n;
+
+
+	res = tee_fs_htree_read_block(ht, bn, b);
+	if (res)
+		return res;
+
+	for (n = 0; n < ARRAY_SIZE(b); n++) {
+		if (b[n] != val_from_bn_n_salt(bn, n, salt)) {
+			DMSG("Unpected b[%zu] %#" PRIx32
+			     "(expected %#" PRIx32 ")",
+			     n, b[n], val_from_bn_n_salt(bn, n, salt));
+			return TEE_ERROR_SECURITY;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result do_range(TEE_Result (*fn)(struct tee_fs_htree **ht,
+					    size_t bn, uint8_t salt),
+			   struct tee_fs_htree **ht, size_t begin,
+			   size_t num_blocks, size_t salt)
+{
+	TEE_Result res = TEE_SUCCESS;
+	size_t n;
+
+	for (n = 0; n < num_blocks; n++) {
+		res = fn(ht, n + begin, salt);
+		CHECK_RES(res, goto out);
+	}
+
+out:
+	return res;
+}
+
+static TEE_Result do_range_backwards(TEE_Result (*fn)(struct tee_fs_htree **ht,
+						      size_t bn, uint8_t salt),
+				     struct tee_fs_htree **ht, size_t begin,
+				     size_t num_blocks, size_t salt)
+{
+	TEE_Result res = TEE_SUCCESS;
+	size_t n;
+
+	for (n = 0; n < num_blocks; n++) {
+		res = fn(ht, num_blocks - 1 - n + begin, salt);
+		CHECK_RES(res, goto out);
+	}
+
+out:
+	return res;
+}
+
+static TEE_Result htree_test_rewrite(struct test_aux *aux, size_t num_blocks,
+				     size_t w_unsync_begin, size_t w_unsync_num)
+{
+	TEE_Result res;
+	struct tee_fs_htree *ht = NULL;
+	size_t salt = 23;
+
+	assert((w_unsync_begin + w_unsync_num) <= num_blocks);
+
+	aux->data_len = 0;
+	memset(aux->data, 0xce, aux->data_alloced);
+
+	res = tee_fs_htree_open(true, &test_htree_ops, aux, &ht);
```

Comment:
could you add few comments about what these tests target/expect ?
A comment above each subtest would be nice.